sycl/source/exception.cpp

```diff
@@ -10,9 +10,105 @@
 #include <CL/sycl/context.hpp>
 #include <CL/sycl/exception.hpp>
 
+#include <cstring>
+
 __SYCL_INLINE_NAMESPACE(cl) {
 namespace sycl {
 
+namespace { // anonymous
+char reserved_for_errorcode[] =
+    "0SYCL1234567812345678"; // 17 (string terminator plus 'SYCL' marker plus
+                             // error code)
+std::error_code sycl121_proxy_errorcode = make_error_code(sycl::errc::invalid);
+} // namespace
+
+exception::exception(std::error_code ec, const char *Msg)
+    : exception(static_cast<context *>(nullptr), ec, Msg) {}
+
+exception::exception(std::error_code ec, const std::string &Msg)
+    : exception(static_cast<context *>(nullptr), ec, Msg) {}
+
+// new SYCL2020 constructors
+exception::exception(std::error_code ec)
+    : exception(static_cast<context *>(nullptr), ec, "") {}
+
+exception::exception(int ev, const std::error_category &ecat,
+                     const std::string &what_arg)
+    : exception(static_cast<context *>(nullptr), {ev, ecat}, what_arg) {}
+
+exception::exception(int ev, const std::error_category &ecat,
+                     const char *what_arg)
+    : exception(static_cast<context *>(nullptr), {ev, ecat},
+                std::string(what_arg)) {}
+
+exception::exception(int ev, const std::error_category &ecat)
+    : exception(static_cast<context *>(nullptr), {ev, ecat}, "") {}
+
+exception::exception(context ctx, std::error_code ec,
+                     const std::string &what_arg)
+    : exception(&ctx, ec, what_arg) {}
+
+exception::exception(context ctx, std::error_code ec, const char *what_arg)
+    : exception(ctx, ec, std::string(what_arg)) {}
+
+exception::exception(context ctx, std::error_code ec)
+    : exception(ctx, ec, "") {}
+
+exception::exception(context ctx, int ev, const std::error_category &ecat,
+                     const char *what_arg)
+    : exception(ctx, {ev, ecat}, std::string(what_arg)) {}
+
+exception::exception(context ctx, int ev, const std::error_category &ecat,
+                     const std::string &what_arg)
+    : exception(ctx, {ev, ecat}, what_arg) {}
+
+exception::exception(context ctx, int ev, const std::error_category &ecat)
+    : exception(ctx, ev, ecat, "") {}
+
+// protected base constructor for all SYCL2020 constructors
+exception::exception(context *ctxPtr, std::error_code ec,
+                     const std::string &what_arg)
+    : MMsg(what_arg + reserved_for_errorcode), MCLErr(PI_INVALID_VALUE),
```

Comment:
At first I was going to suggest following renaming:
```suggestion
    : MMsg(what_arg + reserved_for_errorcode), MPIErr(PI_INVALID_VALUE),
```
but I noticed that there is a method returning this code as `cl_int` value:
```
cl_int exception::get_cl_code() const { return MCLErr; }
```
Is is correct to initialize `MCLErr` with `PI_INVALID_VALUE`? What is the type of `MCLErr`?

Reply:
`MCLErr` is a `cl_int` and is part of the SYCL 1.2.1 exceptions. In SYCL 2020 it is no longer needed, however for compatibility we keep it. We need to specify _some_ value for it, and `PI_INVALID_VALUE` is the most commonly used one in the source base now. This will not affect SYCL 1.2.1 exceptions, because they don't use this protected constructor. The other SYCL 1.2.1 exception classes (like `runtime_error`) pass in their own values. Our use of `PI_INVALID_VALUE` here is just for compatibility, so that `get_cl_code()` will work.  In the future, both `MCLErr` and `get_cl_code()` will be removed.  I will mark `get_cl_code()` as deprecated.